### ISSUE TITLE
How to correctly reply to a invalid PreKeyWhisperMessage

### ISSUE THREAD

**Issue by lovetox** [kalkin/gajim-omemo#98]
Scenario:

```
Client A and Client B have a valid session.

Client A for some reason loses his session record, for example a faulty DB migration
Client A sends an invalid PreKeyWhisperMessage to build a new session
(maybe wrong SignedPreyKeyID, or not matching SignedPreKey,
maybe some error in querying it or converting it from DB)

Client B receives invalid PreKeyWhisperMessage, and does nothing.
Client B send a normal WhisperMessage basing still on the valid session it thinks it has.

Client A sends invalid PreKeyWhisperMessages based on the invalid session it has.
```

how to come out of this state without reseting the DB?

https://conversations.im/xeps/multi-end.html#rules

> As there are no explicit error messages in this protocol, if a client does receive a PreKeyWhisperMessage using an invalid PreKey, they SHOULD respond with a KeyTransportElement, sent in a <message> using a PreKeyWhisperMessage. By building a new session with the original sender this way, the invalid session of the original sender will get overwritten with this newly created, valid session.

So what is a KeyTransportElement, i didnt find it in the axolotl-java-lib

does it just mean, send  a new PreKeyWhisperMessage?
Will this be a new session for Client B, or could we retain the old session in sending the session data in that PreKeyWhisperMessage?
is there an easy reachable function for that case in the lib?

My only idea how to achieve this is, that Client B deletes the session it has with Client A from his DB, so he is forced to build a new one with a PreKeyWhisperMessage, overwritting the faulty session from Client A.
Is there an easier Methode to achieve this?

also i looked at the Signal Android Code, i didnt find anything that looked like they implemented something like that.

@iNPUTmice 
@strb

**Comment by iNPUTmice** [comment]
I haven't thought about an actual solution for this yet but I can clarify what a `KeyTransportElement`  is. Actually thats defined in the Glossary of the XEP as:

> **KeyTransportElement**
> An OMEMO element that does not have a `<payload>`. It contains a fresh encryption key, which can be used for purposes external to this XEP.

So basically it's a OMEMO message without the encrypted body. Technically a `KeyTransportElement` can contain either a PreKeyWhisperMessage or a WhisperMessage. (Even both if you have multiple keys/devices). However in that scenario we are probably talking about a PreKeyWhisperMessage as we want to restore a broken session.

**Comment by di72nn** [comment]
Is the following stack trace related to the issue?

```
(D) gajim.plugin_system.omemo <my jid> => OMEMO MAM msg received
(D) gajim.plugin_system.omemo <my jid> => Received PreKeyWhisperMessage from <their jid>
(E) gajim.c.ged Error while running an even handler: <bound method OmemoPlugin.mam_message_received of <gajim-omemo.OmemoPlugin object at 0x7f61f76cdc90>>
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/gajim/common/ged.py", line 93, in raise_event
    if handler(*args, **kwargs):
  File "/usr/lib/python2.7/site-packages/gajim/plugins/helpers.py", line 107, in wrapper
    result = f(*args, **kwargs)
  File ".../gajim-omemo/__init__.py", line 254, in mam_message_received
    plaintext = state.decrypt_msg(msg_dict)
  File ".../gajim-omemo/omemo/state.py", line 197, in decrypt_msg
    encrypted_key)
  File ".../gajim-omemo/omemo/state.py", line 356, in handlePreKeyWhisperMessage
    key = sessionCipher.decryptPkmsg(preKeyWhisperMessage)
  File "/usr/lib/python2.7/site-packages/axolotl/sessioncipher.py", line 101, in decryptPkmsg
    unsignedPreKeyId = self.sessionBuilder.process(sessionRecord, ciphertext)
  File "/usr/lib/python2.7/site-packages/axolotl/sessionbuilder.py", line 50, in process
    unsignedPreKeyId = self.processV3(sessionRecord, message)
  File "/usr/lib/python2.7/site-packages/axolotl/sessionbuilder.py", line 112, in processV3
    ourSignedPreKey = self.signedPreKeyStore.loadSignedPreKey(message.getSignedPreKeyId()).getKeyPair()
  File ".../gajim-omemo/omemo/liteaxolotlstore.py", line 143, in loadSignedPreKey
    return self.signedPreKeyStore.loadSignedPreKey(signedPreKeyId)
  File ".../gajim-omemo/omemo/litesignedprekeystore.py", line 42, in loadSignedPreKey
    signedPreKeyId)
InvalidKeyIdException: No such signedprekeyrecord! 6233
```

This errors with the same KeyId repeatedly appears in logs.
The problem seems to prevent me from using the OMEMO plugin with one of my contacts.

**Comment by lovetox** [comment]
this should solve itself automatically if you write him a message from that device

if not a log would be useful what happens on his side when you send him a message

**Comment by di72nn** [comment]
I can't send a message because there is no required fingerprint in the fingerprint list. But this is probably somewhat different issue, I'll try to gather more details on it later.

**Comment by lovetox** [comment]
the issue i see here is, you contact doesnt have a devicelist published at his server.

you should try this if he is online at the same time.
as soon as we get a devicelist with keys, and you send him a message, the other problem will be gone.

reasons for a not published devicelist could be: he is not online at the same time, or his server has issues

**Comment by lovetox** [comment]
as this seems to happen like never, i close this.

**Comment by di72nn** [comment]
That's probably not important, but I'll post it just in case: the behavior that I described was fixed in the recent versions (most likely 0.9.7). Thanks for that.
